File: apps/demo/cord-gui/src/main/webapp/app/view/bundle/bundle.js
/*
 * Copyright 2015 Open Networking Laboratory
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

(function () {
    'use strict';

    var urlSuffix = '/rs/bundle';

    var basic = 'basic',
        family = 'family';

    angular.module('cordBundle', [])
        .controller('CordBundleCtrl', ['$log', '$scope', '$resource',
            function ($log, $scope, $resource) {
                var BundleData, resource,
                    getData;
                $scope.page.curr = 'bundle';
                $scope.show = false;

                getData = function (id) {
                    if (!id) { id = ''; }

                    BundleData = $resource($scope.shared.url + urlSuffix + '/' + id);
                    resource = BundleData.get({},
                        // success
                        function () {
                            var current, availId;
                            current = resource.bundle.id;
                            $scope.name = resource.bundle.name;
                            $scope.desc = resource.bundle.desc;
                            $scope.funcs = resource.bundle.functions;

                            availId = (current === basic) ? family : basic;
                            resource.bundles.forEach(function (bundle) {
                                if (bundle.id === availId) {
                                    $scope.available = bundle;
                                }
                            });
                        },
                        // error
                        function () {
                            $log.error('Problem with resource', resource);
                        });
                };

                getData();

                $scope.changeBundle = function (id) {
                    getData(id);
                };

                $scope.showBundles = function () {
                    $scope.show = !$scope.show;
                };

                $log.debug('Cord Bundle Ctrl has been created.');
            }])

        .directive('bundleAvailable', [function () {
            return {
                templateUrl: 'app/view/bundle/available.html'
            };
        }]);
}());

File: apps/demo/cord-gui/src/main/java/org/onosproject/cord/gui/model/Bundle.java
/*
 * Copyright 2015 Open Networking Laboratory
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.onosproject.cord.gui.model;

import com.google.common.collect.ImmutableSet;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;

/**
 * Encapsulates a bundle, including current state.
 */
public class Bundle {
    private final BundleDescriptor bundleDescriptor;
    private final Map<XosFunctionDescriptor, XosFunction> functionMap =
        new HashMap<XosFunctionDescriptor, XosFunction>();

    /**
     * Constructs a new bundle instance.
     *
     * @param bundleDescriptor the descriptor
     */
    public Bundle(BundleDescriptor bundleDescriptor) {
        this.bundleDescriptor = bundleDescriptor;
        initFunctions();
    }

    /**
     * Returns the bundle descriptor.
     *
     * @return the descriptor
     */
    public BundleDescriptor descriptor() {
        return bundleDescriptor;
    }

    /**
     * Returns the set of function instances for this bundle.
     *
     * @return the functions
     */
    public Set<XosFunction> functions() {
        return ImmutableSet.copyOf(functionMap.values());
    }

    /**
     * Creates an initial set of function instances.
     */
    private void initFunctions() {
        for (XosFunctionDescriptor xfd: bundleDescriptor.functions()) {
            functionMap.put(xfd, createFunction(xfd));
        }
    }

    private XosFunction createFunction(XosFunctionDescriptor xfd) {
        XosFunction func;
        switch (xfd) {
            case URL_FILTER:
                func = new UrlFilterFunction();
                break;

            default:
                func = new DefaultXosFunction(xfd);
                break;
        }
        return func;
    }

    /**
     * Returns the function instance for the specified descriptor, or returns
     * null if function is not part of this bundle.
     *
     * @param xfd function descrriptor
     * @return function instance
     */
    public XosFunction findFunction(XosFunctionDescriptor xfd) {
        return functionMap.get(xfd);
    }
}
